core/src/main/java/com/datastax/oss/driver/internal/core/addresstranslation/SubnetAddressTranslator.java

```diff
@@ -0,0 +1,210 @@
+   */
+  public static final String ADDRESS_TRANSLATOR_DEFAULT_ADDRESS =
+      "advanced.address-translator.default-address";
+
+  public static DriverOption ADDRESS_TRANSLATOR_SUBNET_ADDRESSES_OPTION =
+      new DriverOption() {
+        @NonNull
+        @Override
+        public String getPath() {
+          return ADDRESS_TRANSLATOR_SUBNET_ADDRESSES;
+        }
+      };
+  public static DriverOption ADDRESS_TRANSLATOR_DEFAULT_ADDRESS_OPTION =
+      new DriverOption() {
+        @NonNull
+        @Override
+        public String getPath() {
+          return ADDRESS_TRANSLATOR_DEFAULT_ADDRESS;
+        }
+      };
+
+  private static final String DELIMITER = ":";
+  private static final int DEFAULT_PORT = 9042;
+
+  private final List<SubnetAddress> subnetAddresses;
+  private final Optional<InetSocketAddress> defaultAddress;
+  private final String logPrefix;
+
+  public SubnetAddressTranslator(@NonNull DriverContext context) {
+    logPrefix = context.getSessionName();
+    this.subnetAddresses =
+        context.getConfig().getDefaultProfile()
+            .getStringList(ADDRESS_TRANSLATOR_SUBNET_ADDRESSES_OPTION).stream()
+            .map(SubnetAddress::fromString)
+            .collect(Collectors.toList());
+    this.defaultAddress =
+        Optional.ofNullable(
+                context
+                    .getConfig()
+                    .getDefaultProfile()
+                    .getString(ADDRESS_TRANSLATOR_DEFAULT_ADDRESS_OPTION, null))
+            .map(SubnetAddressTranslator::parseAddress);
+    SubnetAddressTranslator.validateSubnetsAreNotOverlapping(this.subnetAddresses);
+  }
+
+  @NonNull
+  @Override
+  public InetSocketAddress translate(@NonNull InetSocketAddress address) {
+    InetSocketAddress translatedAddress = null;
+    for (SubnetAddress subnetAddress : subnetAddresses) {
+      if (subnetAddress.contains(address)) {
+        translatedAddress = subnetAddress.address;
+      }
+    }
+    if (translatedAddress == null && defaultAddress.isPresent()) {
+      translatedAddress = defaultAddress.get();
+    }
+    if (translatedAddress == null) {
+      translatedAddress = address;
+    }
+    LOG.debug("[{}] Resolved {} to {}", logPrefix, address, translatedAddress);
+    return translatedAddress;
+  }
+
+  @Override
+  public void close() {}
+
+  private static InetSocketAddress parseAddress(String address) {
+    String[] addressTuple = address.split(DELIMITER);
+    if (addressTuple.length == 2) {
+      return new InetSocketAddress(addressTuple[0], Integer.parseInt(addressTuple[1]));
+    }
+    if (addressTuple.length == 1) {
+      return new InetSocketAddress(addressTuple[0], DEFAULT_PORT);
+    }
+    throw new IllegalArgumentException("Invalid default address: " + address);
+  }
+
+  private static void validateSubnetsAreNotOverlapping(List<SubnetAddress> subnetAddresses) {
+    for (int i = 0; i < subnetAddresses.size() - 1; i++) {
```

Comment:
This is good to have :+1:.  I imagine without this, there could be some unpredictable behavior if subnets overlap (unclear which endpoint would ultimately be used).